leo_segmentation/data.py

```diff
@@ -1 +1,137 @@
+                all_classes.append(class_name)
+            else:
+                image_class_mapping[class_name] = [fn]
+                all_classes.append(class_name)
+
+        total_files = sum([len(image_class_mapping[i]) for i in image_class_mapping])
+        error_msg = "Not all classes represented in the mapping"
+        assert set(image_class_mapping.keys()) - set(all_classes) == set(), error_msg
+        assert(len(data_dict["embeddings"]) == total_files == len(image_mask_embeddings))
+        return image_class_mapping, image_mask_embeddings
+
+class Datagenerator():
+    """Data Generator class"""
+    def __init__(self, dataset, config, data_type):
+        self.config = config["data_type"][data_type]
+        self.dataset = MetaDataset(dataset, config, data_type)
+        self._data_loader = DataLoader(self.dataset, batch_size=None, shuffle=False,\
+                            num_workers=0, collate_fn=self.collation_fn)
+
+    def get_batch_data(self):
+        return next(iter(self._data_loader))
+
+    def collation_fn(self, data):
+        """
+        Collation function for data loader to generate train and val batch data
+        Args: 
+            data (tuple): contains _all_class_images dict and _image_mask_embeddings
+        
+        Returns:
+            A tuple of pytorch tensors:
+            DIMS = Image or Mask shape
+            - tr_data: (batch_size, num_classes, tr_size, DIMS): training image embeddings
+            - tr_masks: (batch_size, num_classes, tr_size, DIMS): training image masks
+            - val_data: (batch_size, num_classes, val_size, DIMS): validation image embeddings
+            - val_masks: (batch_size, num_classes, val_size, DIMS): validation image masks 
+        """
+        _all_class_images, _image_mask_embeddings = data
+        for i in range(self.config["num_tasks"]):
+            class_list = list(_all_class_images.keys())
+            tr_size = self.config["n_train_per_class"]
+            val_size = self.config["n_val_per_class"]
+            num_classes = self.config["num_classes"]
+            sample_count = (tr_size + val_size)
+            random.shuffle(class_list)
+            shuffled_list = class_list[:num_classes]
+            error_message = f"len(shuffled_list) {len(shuffled_list)} is not num_classes: {num_classes}"
+            assert len(shuffled_list) == num_classes, error_message
+            image_paths = []
+            for _, class_name in enumerate(shuffled_list):
+                all_images = _all_class_images[class_name]
+                all_images = np.random.choice(all_images, sample_count, replace=False)
+                error_message = f"{len(all_images)} == {sample_count} failed"
+                assert len(all_images) == sample_count, error_message
+                image_paths.append(all_images)
+            
+            path_array = np.array(image_paths)
+            embedding_array = np.array([[_image_mask_embeddings[image_path][0] 
```

Comment:
each embedding 2D (32, 20) so ideally this embedding array of batch_size (input_task) = 12, tr + val = 8, should result in (12,8,32,20) but the above logic is resulting in (12, 8, 8, 32, 20) am I miss understanding something here?

Reply:
_Good observation_ . In the doc-string I wrote this:
`
Returns:
            A tuple of pytorch tensors:
            DIMS = Image or Mask shape
            - tr_data: (batch_size, num_classes, tr_size, DIMS): training image embeddings
            - tr_masks: (batch_size, num_classes, tr_size, DIMS): training image masks
            - val_data: (batch_size, num_classes, val_size, DIMS): validation image embeddings
            - val_masks: (batch_size, num_classes, val_size, DIMS): validation image masks 
  `

Please confirm it does that. Since we are going to train per class, there is also a dimension for class. We were discussing about averaging codes per class for the relational network.

Reply:
Yes for relation network we might need the extra dimension while concatenating each encoder output with every other, so I can tweak it as per the model needs, I will modify the above logic to get (12, 8, 32, 20) to pass as input to the encoder and in the later step in relation network, I can use the above logic of yours.